Userland/Test.cpp

```diff
@@ -0,0 +1,540 @@
+    } else if (arg == "(") {
+        optind++;
+        auto command = parse_complex_expression(argv);
+        if (command && argv[optind] && StringView(argv[++optind]) == ")")
+            return command;
+        fatal_error("Unmatched \033[1m(");
+    } else if (arg == "!") {
+        optind++;
+        auto command = parse_complex_expression(argv);
+        if (command)
+            return make<Not>(command.release_nonnull());
+        fatal_error("Expected an expression after \033[1m!");
+    } else if (arg == "-e") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-e expects a file name");
+        return make<Exists>(name);
+    } else if (arg == "-b") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-b expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::BlockDevice);
+    } else if (arg == "-c") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-c expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::CharacterDevice);
+    } else if (arg == "-d") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-d expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Directory);
+    } else if (arg == "-f") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-f expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Regular);
+    } else if (arg == "-h" || arg == "-L") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-h/-L expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::SymbolicLink);
+    } else if (arg == "-p") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-p expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::FIFO);
+    } else if (arg == "-S") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-S expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Socket);
+    } else if (arg == "-r") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-r expects a file name");
+        return make<HasPermission>(name, HasPermission::Read);
+    } else if (arg == "-w") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-w expects a file name");
+        return make<HasPermission>(name, HasPermission::Write);
+    } else if (arg == "-x") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-x expects a file name");
+        return make<HasPermission>(name, HasPermission::Execute);
+    } else if (arg == "-a") {
+        return nullptr;
```

Comment:
This here requires a comment for why we return `nullptr` 🙂


I suppose it's that "and" is not a simple command, so we're returning back to our caller to handle it.